Sources/PackageSyntax/PackageEditor.swift

```diff
@@ -0,0 +1,469 @@
+
+        try diagnoseUnsupportedToolsVersions(manifest: loadedManifest)
+
+        let containsDependency = loadedManifest.dependencies.contains {
+            return PackageIdentity(url: URL(string: url)!) == $0.identity
+        }
+        guard !containsDependency else {
+            context.diagnosticsEngine.emit(.packageDependencyAlreadyExists(url: url,
+                                                                           packageName: loadedManifest.displayName))
+            throw Diagnostics.fatalError
+        }
+
+        // If the input URL is a path, force the requirement to be a local package.
+        if URL(string: url)?.scheme == nil {
+            guard requirement == nil || requirement == .localPackage else {
+                context.diagnosticsEngine.emit(.nonLocalRequirementSpecifiedForLocalPath(path: url))
+                throw Diagnostics.fatalError
+            }
+            requirement = .localPackage
+        }
+
+        // Load the dependency manifest depending on the inputs.
+        let dependencyManifest: Manifest
+        if requirement == .localPackage {
+            let path = AbsolutePath(url, relativeTo: fs.currentWorkingDirectory!)
+            dependencyManifest = try context.loadManifest(at: path)
+            requirement = .localPackage
+        } else {
+            // Otherwise, first lookup the dependency.
+            let id = PackageIdentity(url: URL(string: url)!)
+            let spec = RepositorySpecifier(url: URL(string: url)!)
+            let handle = try tsc_await{
+                let observability = ObservabilitySystem { _, _ in }
+                context.repositoryManager.lookup(package: id, repository: spec, skipUpdate: true, observabilityScope: observability.topScope, delegateQueue: .global(qos: .userInitiated), callbackQueue: .global(qos: .userInitiated), completion: $0)
+            }
+            let repo = try handle.open()
+
+            // Compute the requirement.
+            if let inputRequirement = requirement {
+                requirement = inputRequirement
+            } else {
+                // Use the latest version or the main/master branch.
+                let versions = try repo.getTags().compactMap{ Version($0) }
+                let latestVersion = versions.filter({ $0.prereleaseIdentifiers.isEmpty }).max() ?? versions.max()
+                let mainExists = (try? repo.resolveRevision(identifier: "main")) != nil
+                requirement = latestVersion.map{ PackageDependencyRequirement.upToNextMajor($0.description) } ??
+                    (mainExists ? PackageDependencyRequirement.branch("main") : PackageDependencyRequirement.branch("master"))
+            }
+
+            // Load the manifest.
+            let revision = try repo.resolveRevision(identifier: requirement!.ref!)
+            let repoFS = try repo.openFileView(revision: revision)
+            dependencyManifest = try context.loadManifest(at: .root, fs: repoFS)
+        }
```

Comment:
this ^^ does not seem to be correct, it is also bit of repetition of the parsing code from the manifest parser. we should find way to consolidate so its more correct. it also needs to be changes to address registry based dependencies